### PR TITLE
flexible formats for force constants in HessianJob

### PR DESCRIPTION
**Pull request by samwaseda**
Following @jan-janssen 's comment, replace the core part of `EinstenExampleJob` by `HessianJob`

### CODE REVIEW

**Comment by jan-janssen** [comment]
I am not so much a fan of the if return pattern, should not it be just one if elif, .. else block, to be more readable.

**Comment by samwaseda** [comment]
> I am not so much a fan of the if return pattern, should not it be just one if elif, .. else block, to be more readable.

I'm not either, but I couldn't come up with a better solution

**Comment by samwaseda** [comment]
This class is now ready for review (sorry for setting review so early)

File: pyiron/thermodynamics/hessian.py [hunk 1]
```diff
@@ -18,40 +18,117 @@
 
 class HessianJob(GenericInteractive):
     def __init__(self, project, job_name):
-        super(GenericInteractive, self).__init__(project, job_name)
+        super(HessianJob, self).__init__(project, job_name)
         self.__version__ = "0.0.1"
         self.__name__ = "HessianJob"
-        self.interactive_cache = {
```
Comment (jan-janssen) [inline pyiron/thermodynamics/hessian.py:24]: Why is this removed?

**Comment by jan-janssen** [comment]
I have the feeling this pull request breaks the previous examples.

File: pyiron/thermodynamics/hessian.py [hunk 2]
```diff
@@ -18,40 +18,117 @@
...
-            "forces": [],
-            "positions": [],
-            "energy_pot": []
-        }
+        self.server.run_mode.interactive = True
         self._force_constants = None
         self._reference_structure = None
-        self._next_positions = None
+        self._energy_pot = 0
```
Comment (jan-janssen) [inline pyiron/thermodynamics/hessian.py:27]: I do not understand this part.

File: pyiron/thermodynamics/hessian.py (hunk 1 above)
Comment (samwaseda) [reply pyiron/thermodynamics/hessian.py:24]: Because since the introduction of `interactive_output_functions`, `interactive_cache` is not necessary anymore.

File: pyiron/thermodynamics/hessian.py [hunk 3]
```diff
@@ -18,40 +18,117 @@
...
-        super(GenericInteractive, self).__init__(project, job_name)
+        super(HessianJob, self).__init__(project, job_name)
         self.__version__ = "0.0.1"
         self.__name__ = "HessianJob"
-        self.interactive_cache = {
-            "forces": [],
-            "positions": [],
-            "energy_pot": []
-        }
+        self.server.run_mode.interactive = True
```
Comment (jan-janssen) [inline pyiron/thermodynamics/hessian.py:24]: This part is also confusing for me.

File: pyiron/thermodynamics/hessian.py [hunk 4]
```diff
@@ -18,40 +18,117 @@
...
+            self.structure = structure.copy()
+
+    def validate_ready_to_run(self):
+        super(HessianJob, self).validate_ready_to_run()
+        if self._force_constants is None:
+            raise AssertionError('set force constants by set_force_constants before run')
+        if self._reference_structure is None:
+            raise AssertionError('set reference structure by set_reference_structure before run')
+
+    def interactive_forces_getter(self):
```
Comment (jan-janssen) [inline pyiron/thermodynamics/hessian.py:74]: I do not think this is necessary if we use the interactive_cache.

Comment (samwaseda) [reply pyiron/thermodynamics/hessian.py:74]: That's true. Either way is possible. The slight advantage of defining getters is that the user can also remove some of them. But in this job probably this plays no significant role.

**Comment by samwaseda** [comment]
> I have the feeling this pull request breaks the previous examples.

Which ones?

File: pyiron/thermodynamics/hessian.py (hunk 2 above)
Comment (samwaseda) [reply pyiron/thermodynamics/hessian.py:27]: You need to put more info for me to understand what you don't understand.

File: pyiron/thermodynamics/hessian.py (hunk 3 above)
Comment (samwaseda) [reply pyiron/thermodynamics/hessian.py:24]: Here as well.

File: pyiron/thermodynamics/hessian.py (hunk 2 above)
Comment (jan-janssen) [reply pyiron/thermodynamics/hessian.py:27]: Why is the` interactive_cache` removed and replaced by individual variables?

File: pyiron/thermodynamics/hessian.py (hunk 4 above)
Comment (jan-janssen) [reply pyiron/thermodynamics/hessian.py:74]: No from my perspective that is a general design decision to implement the cache for all jobs.

File: pyiron/thermodynamics/hessian.py (hunk 2 above)
Comment (samwaseda) [reply pyiron/thermodynamics/hessian.py:27]: Because in principle we don't touch `interactive_cache` anymore to determine which quantities should be stored. Instead of creating individual variables I could also calculate the values each time, but to make it faster I did it this way.

**Comment by jan-janssen** [comment]
https://github.com/pyiron/pyiron-meeting-mpie/blob/2a9c7423d4cf68f8ede6c5460e0058fce0e14ac0/meetings/2020-02-21-pyiron-meeting/jan-janssen/thermoint.ipynb

File: pyiron/thermodynamics/hessian.py (hunk 2 above)
Comment (jan-janssen) [reply pyiron/thermodynamics/hessian.py:27]: As I said for me merging this pull request without discussing it in detail breaks examples and it breaks the design we implemented for all other interactive jobs. I am open to discuss both but then we should also adjust all other jobs.

**Comment by samwaseda** [comment]
> https://github.com/pyiron/pyiron-meeting-mpie/blob/2a9c7423d4cf68f8ede6c5460e0058fce0e14ac0/meetings/2020-02-21-pyiron-meeting/jan-janssen/thermoint.ipynb

For me this already fails when I try to load `PhonopyMaster`

**Comment by samwaseda** [comment]
ok it works if it's changed to `PhonopyJob`

File: pyiron/thermodynamics/hessian.py (hunk 2 above)
Comment (samwaseda) [reply pyiron/thermodynamics/hessian.py:27]: Which design? `interactive_cache` is now not touched in any of the major codes (LAMMPS, SPHInX, VASP)

Comment (samwaseda) [reply pyiron/thermodynamics/hessian.py:27]: And I find it extremely unfair to talk about breaking examples if there's no unit test at all and no DocString explaining how the functions are supposed to be used.
